### ISSUE TITLE
Add Loop syntax for Calculations

### ISSUE THREAD

**Issue by joeflack4** [getodk/xforms-spec#85]
Original thread: https://groups.google.com/forum/#!topic/javarosa-developers/6Z1fSlJdb7Q):
_________________________

My organization uses ODK and would like to add a new 'module' of questions to some of our current surveys. We are using Xforms. In particular, we are writing our forms in Excel (XLS forms), and converting.

A major thing desired has been the ability to have dynamic constraints. We are struggling to think of a way out of our problem without some sort of ability to declare a 'for loop' style iteration.

Basically we want a repeat, where in each iteration we ask about the respondent's status during a particular period of time. Each repeat is a group, where there are fields for 'start date', 'end date', and 'status'.

What we would like is the ability to make sure that none of these periods are in conflict. The end date of one period cannot occur after the start date of another period. They cannot overlap. We want that when a respondent were to need to go back and change a previous response, it would be able to *iterate through the start/end dates for each response within the repeat* and check each one to make sure there is no conflict.

We think it might be nice to have syntax as would follow below.

*Desired Syntax*
`for_each([repeatName], [groupName], [fieldName], [representativeReferenceVariable], [iterator], [conditions])`

*Example*
This would be the constraint in the 'start_date' field of a repeat instance of a 'pregnancy_history' question group.

_Single Line_
`for_each(pregnancy_history_repeat, pregnancy_history_group, end_date, myRef, i, (${myRef} < .))`

_Extended_
```for_each(
pregnancy_history_repeat, pregnancy_history_group, end_date, 
myRef, 
i, 
(${myRef} < .)
)```

So in this example above, we would be identifying our reference variable as the 'end_date' field variable within a group, within a repeat. We would then be creating an in-line reference variable 'myRef', and declaring an iterator variable ('i') if for any reason needed. For each iteration, we're comparing the 'end_date' field value of that iteration to the value entered in the current field (represented in XlsForm/Xform syntax as a ' . '), ie. the 'start_date' field. (edited)

**Comment by MartijnR** [comment]
Thanks for posting this!

I think adding support for the native [XPath 1.0 preceding-sibling axis](https://www.w3.org/TR/xpath/#axes) would be the way to go. This way you can put a constraint on each start that compares with an end value in the directly-preceding repeat.

(P.S. if you want to test this, the axis is already unofficially supported in Enketo)

**Comment by MartijnR** [comment]
actually the same approach might already work with position() using `. > /path/to/repeat[position() = (position(current()/..) - 1)]/end` or something similar.

**Comment by joeflack4** [comment]
Hey there. Thanks for your helpful input.

It looks like you're right that this could be part of what we're looking for. If I have a repeat group of, let's say, birthdays, and I want to ensure no dates entered can be the same, I could constrain for all children in the repeat node using preceding- and following- sibling. We would want to constrain for following siblings in the event that a previous entry needed to be changing.

If it wasn't for that last edge case, I think that would work. I could have a calculate for each repeat group that appended the previous groups birthday(s), essentially building a list with each iteration. But alas, the edge case of looking back and forward for all siblings.

We would also want to loop through all children of a repeat from outside the repeat.

Any thoughts on this?

**Comment by MartijnR** [comment]
That birthday problem is interesting, because you have to avoid a circular reference. That's why `not(selected(join(' ', /path/to/repeat/birthday), .) )` won't work. (I think the for_each function wouldn't avoid that).

I think the `following` and `preceding` axes (not -sibling) are the way to do this in combination with join() and selected(). Another way would  be to use following and preceding axes with a predicate that searches for the current value (and counts the hits), e.g. something like (likely contains errors - and may also very well be utterly wrong too):

```
count(../preceding::repeat/birthday[text() = current()/.] | ../following::repeat/birthday[text() = current()/.]) = 0
```

So if that would work we'd be talking about adding support for at least `preceding-sibling`, `preceding` and `following` axes, and then perhaps later a smart shortcut function that makes it easier to work with them.

P.S. It's hard to test this because it won't pass ODK Validate, but if you publish your XForms on a public server, I can you hook it up with odk.enke.to. Just let me know the domain or IP address.

**Comment by joeflack4** [comment]
I agree with you. A loop alone without code to handle self/circular reference would be a problem. But certainly it would seem there are some ways we could work around this, as you propose.

Checking all siblings is definitely a use case we would like to solve. But just as important for us is for a repeat group to reference another repeat group. Some of our team prefer survey workflows that, for example, ask a series of questions for members of a household, and then later on in the survey would like to ask another series of questions for those same members, with some validations included.

By the way, what kinds of validations is Enketo able to validate that ODK Validate normally wouldn't catch? I know that some errors slip by us and sneak into our surveys, despite our best efforts.

**Comment by MartijnR** [comment]
I think you covered 3 use cases. Some will require new XPath axes to be supported. Others are already supported with `/path/to/repeat[2]/node` syntax (or the indexed-repeat() shortcut), or clever use of join() with a space-separator and selected().

> By the way, what kinds of validations is Enketo able to validate that ODK Validate normally wouldn't catch? I know that some errors slip by us and sneak into our surveys, despite our best efforts.

There is no validator around Enketo's engine. So using any XPath 1.0 features that are not yet supported in JavaRosa requires bypassing validation completely. ODK Validate is the 1 tool that users of all 3 unique form engines rely on. (PS. Enketo supports those features accidentally, because its engine is using an off-the-shelf XPath 1.0 evaluator and extended it). [Update: Enketo has a validator now: https://github.com/enketo/enketo-validate)]

For what this spec concerns, I'm in favor of adding the native XPath 1.0 `preceding-sibling`, `preceding` and `following` XPath axes. I think any shortcut functions should be proposed afterwards, when the core can handle them in native XPath.
